projects/exui-common-lib/src/lib/components/generic-filter/generic-filter.component.ts

```diff
@@ -1,37 +1,71 @@
-import { Component, Input, OnInit, ViewEncapsulation } from '@angular/core';
-import { FilterConfig, FilterSetting } from '../../models/filter.model';
-import { FilterService } from '../../services/filter/filter.service';
+import {ChangeDetectionStrategy, Component, Input, ViewEncapsulation} from '@angular/core';
+import {FilterConfig, FilterSetting} from '../../models';
+import {FilterService} from './../../services/filter/filter.service';
 
 @Component({
   selector: 'xuilib-generic-filter',
   templateUrl: 'generic-filter.component.html',
   styleUrls: ['generic-filter.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
   encapsulation: ViewEncapsulation.None
 })
-export class GenericFilterComponent implements OnInit {
-
-  @Input() public config: FilterConfig;
-  @Input() public settings?: FilterSetting;
+export class GenericFilterComponent {
   private selected: { name: string, value: string[] }[] = [];
 
   constructor(private readonly filterService: FilterService) {
   }
 
-  public ngOnInit(): void {
-    if (!this.settings) {
+  // tslint:disable-next-line:variable-name
+  private _config: FilterConfig;
+
+  public get config(): FilterConfig {
+    return this._config;
+  }
+
+  @Input()
+  public set config(value: FilterConfig) {
+    this._config = {
+      ...value,
+      fields: value.fields.map(field => ({
+        ...field,
+        displayMinSelectedError: false,
+        displayMaxSelectedError: false,
+      }))
+    };
+    this._config = value;
+  }
+
+  public hasErrors = false;
+
+  // tslint:disable-next-line:variable-name
+  private _settings?: FilterSetting;
+
+  public get settings(): FilterSetting {
+    return this._settings;
+  }
+
+  @Input()
+  public set settings(value: FilterSetting) {
+    if (!value) {
       this.getSettings();
     }
-    if (this.settings && this.settings.fields) {
-      this.selected = this.settings.fields;
+    if (value && value.fields) {
+      this.mergeDefaultFields(value);
     }
+    this._settings = value;
   }
 
   public applyFilter(): void {
-    this.filterService.persist(this.settings, this.config.persistence);
+    if (!this.checkFieldsConstraints(this.selected)) {
+      this.filterService.givenErrors.next(null);
+      this.filterService.persist(this.settings, this.config.persistence);
+    }
   }
 
   public cancelFilter(): void {
-    this.getSettings();
+    this._settings.fields = JSON.parse(JSON.stringify(this.config.cancelSetting.fields));
```

Comment:
JSON.parse(JSON.stringify(this.config.cancelSetting.fields)) is expensive operation, we do this twice in line 66/67.
I don't like to use JSON.parse(JSON.stringify(*), see https://medium.com/@pmzubar/why-json-parse-json-stringify-is-a-bad-practice-to-clone-an-object-in-javascript-b28ac5e36521, 
but if we have to use it, just call it once and assign it to a const, and avoid call it twice.

Reply:
Done